api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -181,11 +181,34 @@ public static Map<Integer, String> indexQuotedNameById(
     return indexer.byId();
   }
 
+  /**
+   * Creates a mapping from lower-case field names to their corresponding field IDs.
+   *
+   * <p>This method iterates over the fields of the provided struct and maps each field's name
+   * (converted to lower-case) to its ID. If two fields have the same lower-case name, an
+   * `IllegalArgumentException` is thrown.
+   *
+   * @param struct the struct type whose fields are to be indexed
+   * @return a map where the keys are lower-case field names and the values are field IDs
+   * @throws IllegalArgumentException if two fields have the same lower-case name
+   */
   public static Map<String, Integer> indexByLowerCaseName(Types.StructType struct) {
     Map<String, Integer> indexByLowerCaseName = Maps.newHashMap();
     indexByName(struct)
         .forEach(
-            (name, integer) -> indexByLowerCaseName.put(name.toLowerCase(Locale.ROOT), integer));
+            (name, integer) -> {
+              String key = name.toLowerCase(Locale.ROOT);
+              Integer prevValue = indexByLowerCaseName.get(key);
```

Comment:
I think this check could be much shorter:

```java
String key = name.toLowerCase(Locale.ROOT);
Integer existingId = indexByLowerCaseName.put(key);
Preconditions.checkArgument(
    existingId == null || existingId = fieldId,
    "Cannot build lower case index: %s and %s collide",
    idToName.get(fieldId),
    idToName.get(existingId));
```

Reply:
`Preconditions.checkArgument` throws IllegalArgumentException. This conflicts with your other comment that suggests throwing IllegalStateException. What's the path forward?